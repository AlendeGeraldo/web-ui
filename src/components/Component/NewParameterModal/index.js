/**
 * Component responsible for:
 * - Structuring the advanced new component parameter layout
 * - Add parameter
 */
import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Form, Input, message, Modal, Select, Switch } from 'antd';
import { addParam } from '../../../store/actions/componentActions';

const { Option } = Select;
const { TextArea } = Input;

const NewParameterModal = (props) => {
  const { details, form, visible, onCancel, onAddParam } = props;
  const { uuid, parameters } = details;
  const { getFieldDecorator, getFieldsError } = form;

  /**
   * Function used to check if form has errors
   * @param {string[]} fieldsError
   */
  const hasErrors = (fieldsError) => {
    return Object.keys(fieldsError).some((field) => fieldsError[field]);
  };

  /**
   * Function to handle modal cancel
   */
  const handleCancel = () => {
    form.resetFields();
    onCancel();
  };

  /**
   * Function to handle form submit
   * @param {Event} e
   */
  const handleSubmit = (e) => {
    e.preventDefault();
    form.validateFields((err, values) => {
      if (!err) {
        const newParameter = {
          name: values.name,
          type: values.type,
          required: values.required,
          default: values.defaultValue,
          details: values.details,
        };

        onAddParam(uuid, parameters, newParameter).then((response) => {
          if (response) {
            message.success(
              `Par창metro ${newParameter.name} adicionado com sucesso`
            );
            handleCancel();
          }
        });
      }
    });
  };

  return (
    <Modal
      visible={visible}
      title='Novo Par창metro'
      okText='Criar'
      cancelText='Cancelar'
      onCancel={handleCancel}
      onOk={handleSubmit}
      okButtonProps={{ disabled: hasErrors(getFieldsError()) }}
    >
      <Form layout='vertical'>
        <Form.Item label='Nome: '>
          {getFieldDecorator('name', {
            rules: [{ required: true, message: ' ' }],
          })(<Input />)}
        </Form.Item>
        <Form.Item label='Tipo: '>
          {getFieldDecorator('type', {
            rules: [{ required: true, message: ' ' }],
          })(
            <Select>
              <Option value='float'>Float</Option>
              <Option value='int'>Int</Option>
              <Option value='string'>String</Option>
            </Select>
          )}
        </Form.Item>
        <Form.Item label='Valor padr찾o: '>
          {getFieldDecorator('defaultValue')(<Input />)}
        </Form.Item>
        <Form.Item label='Detalhes: '>
          {getFieldDecorator('details')(<TextArea />)}
        </Form.Item>
        <Form.Item label='Obrigat처rio: '>
          {getFieldDecorator('required', {
            valuePropName: 'checked',
            initialValue: true,
          })(<Switch />)}
        </Form.Item>
      </Form>
    </Modal>
  );
};

NewParameterModal.propTypes = {
  visible: PropTypes.bool.isRequired,
  onCancel: PropTypes.func.isRequired,
  form: PropTypes.shape({
    getFieldDecorator: PropTypes.func,
    getFieldsError: PropTypes.func,
  }).isRequired,
};

const mapStateToProps = (state) => {
  return {
    details: state.component.details,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    onAddParam: (id, parameters, newParameter) => {
      return dispatch(addParam(id, parameters, newParameter));
    },
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Form.create({ name: 'new-parameter-modal' })(NewParameterModal));
