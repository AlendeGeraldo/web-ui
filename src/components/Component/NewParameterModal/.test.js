import React from 'react';

import { shallow, mount } from 'enzyme'

import { act } from 'react-dom/test-utils';

import { Modal, Form, Input, Select, Switch } from 'antd';

import NewParameterModal from '.';

const teste = () => {}

describe('NewParameterModal component', () => {

  it('is expected render without crashing', () => {
    shallow(
      <NewParameterModal 
        visible={true}
        onCancel={teste}
        onCreate={teste}/>
      );
  });

  it('is expected render html correctly', () => {
    const wrapper = shallow(
      <NewParameterModal
        visible={true}
        onCancel={teste}
        onCreate={teste}/>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('is expected to exist Modal', () => {
    const wrapper = mount(
      <NewParameterModal
        visible={true}
        onCancel={teste}
        onCreate={teste}/>
    );

    expect(
      wrapper
        .find(Modal)
        .exists()
    ).toBeTruthy();
  });

  it('is expected to exist Form', () => {
    const wrapper = mount(
      <NewParameterModal
        visible={true}
        onCancel={teste}
        onCreate={teste}/>
    );

    expect(
      wrapper
        .find(Form)
        .exists()
    ).toBeTruthy();
  });
});
