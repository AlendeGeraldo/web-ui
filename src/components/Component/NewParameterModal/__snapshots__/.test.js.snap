// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewParameterModal component is expected render html correctly 1`] = `
<NewParameterModal
  form={
    Object {
      "getFieldDecorator": [Function],
      "getFieldError": [Function],
      "getFieldInstance": [Function],
      "getFieldProps": [Function],
      "getFieldValue": [Function],
      "getFieldsError": [Function],
      "getFieldsValue": [Function],
      "isFieldTouched": [Function],
      "isFieldValidating": [Function],
      "isFieldsTouched": [Function],
      "isFieldsValidating": [Function],
      "isSubmitting": [Function],
      "resetFields": [Function],
      "setFields": [Function],
      "setFieldsInitialValue": [Function],
      "setFieldsValue": [Function],
      "submit": [Function],
      "validateFields": [Function],
      "validateFieldsAndScroll": [Function],
    }
  }
  onCancel={[Function]}
  onCreate={[Function]}
  visible={true}
/>
`;
